torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -405,6 +413,51 @@ std::string TensorPipeAgent::createUniqueShmAddr() {
 }
 #endif
 
+std::string TensorPipeAgent::getDefaultIPAddress() {
+  char hostname[HOST_NAME_MAX];
+  int rv = gethostname(hostname, HOST_NAME_MAX);
+  TORCH_INTERNAL_ASSERT(rv == 0, "Unable to get local hostname");
+
+  struct addrinfo hints, *servinfo, *p;
+  memset(&hints, 0, sizeof hints);
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  rv = getaddrinfo(hostname, nullptr, &hints, &servinfo);
+  TORCH_INTERNAL_ASSERT(rv == 0, "Get address info error: ", gai_strerror(rv));
+
+  // Loop through all the results and pick up the first we can bind.
+  for (p = servinfo; p != NULL; p = p->ai_next) {
+    int fd = socket(p->ai_family, p->ai_socktype, p->ai_protocol);
+    if (fd == -1) {
+      continue;
+    }
+    int bind_rv = bind(fd, p->ai_addr, p->ai_addrlen);
+    if (bind_rv == -1) {
+      continue;
+    }
+
+    if (p->ai_family == AF_INET6) {
+      std::string ipv6(INET6_ADDRSTRLEN, '\0');
+      struct sockaddr_in6* h = (struct sockaddr_in6*)p->ai_addr;
+      inet_ntop(AF_INET6, &h->sin6_addr, (char*)ipv6.data(), INET6_ADDRSTRLEN);
+      freeaddrinfo(servinfo);
+      return ipv6;
```

Comment:
You cannot return early here or you will skip the `freeaddrinfo(servinfo);` call below and leak memory. Perhaps a RAII wrapper around it would be best here, so that we can also deal with exceptions properly.

Reply:
On line 433 it's calling freeaddrinfo(servinfo)

Reply:
Yes but if you *return* you're skipping all the rest of the function and thus you're also skipping the `freeaddrinfo`.